wazuh/wazuh-qa
# Vulnerability Detector refactor - Test scans

|Related issue|
|-------------|
| closes #3309 |

## Description

The purpose of these tests is to check the vulnerability detector scan and the scan results.

### Test cases per module

<details>

<br>

- `test_scan.py`:
  - `test_scan()`: Checks if the scan is triggered in some scenarios unrelated to specific events.
    - Test cases:
      - Check that the scan is triggered only for the agents that meet the requirements
      - Check that the scan is triggered even if the agent information is old
      - Check that the scan is triggered in all agents when the vulnerability affects all of them
  - `test_scan_on_firs_start()`: Checks if the scan is triggered and completed correctly when the manager has started for the first time.
    - Test cases:
      - Check that a scan starts when the manager starts for the first time
  - `test_scan_on_event()`: Checks if the scan is triggered when specific events are generated in the endpoints.
    - Test cases:
      - Check that a new scan is triggered after installing a package
      - Check that a new scan is triggered after updating a package
      - Check that a new scan is triggered after removing a package
      - Check that a new scan is triggered after updating the feed
      - Check that the scan is triggered when an agent is added
  - `test_scan_not_triggered()`: Checks if the scan does not start when certain criteria are met.
    - Test cases:
      - Check that the scan is not triggered if the feed update has failed
      - Check that the scan is not triggered when there is no information in the agent database
      - Check that the scan is not triggered when the agent database does not exist
      - Check that the scan is not triggered when the vulnerability does not affect any agent
      - Check that the scan is not triggered when an invalid configuration was applied
- `test_scan_result.py`:
  - `test_scan_result()`: Checks if the result is published and if its content is correct under specific scenarios.
    - Test cases:
      - Check that the vulnerable package is marked as solved after upgrading
      - Check that the non-vulnerable package is marked as affected after upgrading
      - Check that a non-vulnerable package does not appear in the scan results
      - Check that a removed package does not appear in the scan results
  - `test_scan_result_reprocessing()`: Checks if the result reprocessing works as expected. _When the result is available a Post Scan Action (PSA) is triggered to publish the result in each subscriber, if the PSA failed and the subscriber is a priority subscriber then the publish is queued until it can be processed again._

</details>

### Added

1. `test_scan.py`: This module checks if a scan is started or not under specific circumstances.
2. `test_scan_result.py`: This module checks if the result of the scan is published or not under specific circumstances and if its content is correct.

### Updated

- `agent_db`: added a function to delete hotfixes.
- Vulnerability Detector module (framework): added a vendors list, and a list with the name of the OS per vendor.

---

## Testing performed


| Tester             | Test path | Jenkins | Local  | OS | Commit | Notes                |
|--------------------|-----------|---------|--------|-----|--------|----------------------|
| @user (Developer)  |           | :no_entry_sign: :no_entry_sign: :no_entry_sign: | :no_entry_sign: :no_entry_sign: :no_entry_sign: |         |         | Nothing to highlight |
| @user (Reviewer)   |           | :no_entry_sign: :no_entry_sign: :no_entry_sign: | :no_entry_sign: :no_entry_sign: :no_entry_sign:  |        |         | Nothing to highlight |
